### ISSUE TITLE
Pytorch bert large pretraining hits endless loop of gradient overflow after ~ 350 steps

### ISSUE THREAD

**Issue by David-Levinthal** [NVIDIA/DeepLearningExamples#252]
This was done on a dgx2, using the Nvidia docker image built with the scripts (build.sh, launch.sh). The data set was built per the instructions in the readme using create_datasets_from_start.sh per instructions. This took a while.
When running the scripts/run_pretraining.sh script (in docker env) but modifying it to uns 16 GPUs I hit an endless loop of loss output that looks like:
Step:338 Average Loss = nan Step Loss = nan LR 0.006
Iteration:  51%|\u2588\u2588\u2588\u2588\u2588\u258f    | 5683/11037 [31:18<28:47,  3.10it/s]Rank 0 :: Gradient overflow.  Skipping step, reducing loss scale to 3.637978807091713e-12
Step:338 Average Loss = nan Step Loss = nan LR 0.006
Iteration:  53%|\u2588\u2588\u2588\u2588\u2588\u258e    | 5811/11037 [32:00<28:06,  3.10it/s]Rank 0 :: Gradient overflow.  Skipping step, reducing loss scale to 1.8189894035458565e-12
Step:338 Average Loss = nan Step Loss = nan LR 0.006
Iteration:  54%|\u2588\u2588\u2588\u2588\u2588\u258d    | 5939/11037 [32:42<27:22,  3.10it/s]Rank 0 :: Gradient overflow.  Skipping step, reducing loss scale to 9.094947017729282e-13
Step:338 Average Loss = nan Step Loss = nan LR 0.006
Iteration:  54%|\u2588\u2588\u2588\u2588\u2588\u258d    | 5981/11037 [32:57<27:08,  3.10it/s]
the log files are in the attached tarball along with the run_pretraining.sh script as used for the second log (bert_lamb_np16_p1p2_2.log). For the bert_lamb_np16_p1p2.log the run_pretraining.sh was slightly different from what is in the tarball  in that I lowered 
warmup_proportion=${5:-"0.2843"}
to
warmup_proportion=${5:-"0.128"}
thinking that would save a bit of time.
Thus for the second run the only differences from the default run_pretraining script are below
where the DeepLearningExamples tree was cloned just before running this diff
~/pyt_nan$ diff run_pretraining.sh ../DeepLearningExamples/PyTorch/LanguageModeling/BERT/scripts/run_pretraining.sh

20c20
< num_gpus=${4:-16}
---
> num_gpus=${4:-8}
22c22
< train_steps=${6:-3520}
---
> train_steps=${6:-7038}
37c37
< train_steps_phase2=${20:-800}
---
> train_steps_phase2=${20:-1563}

note the dle_pytorch_nan.zip is a renamed dle_pytorch_nan.tgz  as I could not upload a tgz file :-)

[dle_pytorch_nan.zip](https://github.com/NVIDIA/DeepLearningExamples/files/3719492/dle_pytorch_nan.zip)

**Comment by David-Levinthal** [comment]
This was due to not reducing the gradient_accumulation_steps=${11:-128}  to 64 when going to 16 GPUs. This is stated in the Readme in one place. It might be good to add a run_pretraining_numgpu16.sh script so people can't miss the issue  :-)
feel free to close
